deals/deals.go

```diff
@@ -225,6 +234,195 @@ func (m *Module) Watch(ctx context.Context, proposals []cid.Cid) (<-chan DealInf
+	if err != nil {
+		return nil, err
+	}
+	pending, err := m.store.getPendingDeals()
+	if err != nil {
+		return nil, err
+	}
+	ret = append(ret, pending...)
+	sort.Slice(ret, func(i, j int) bool {
+		if ret[i].Pending || ret[j].Pending {
+			return ret[i].Time < ret[j].Time
+		}
+		if ret[i].DealInfo.ActivationEpoch < ret[j].DealInfo.ActivationEpoch {
+			return true
+		}
+		if ret[i].DealInfo.ActivationEpoch > ret[j].DealInfo.ActivationEpoch {
+			return false
+		}
+		return ret[i].Time < ret[j].Time
+	})
+	return ret, nil
+}
+
+// RetrievalRecords returns a list of all retrievals.
+func (m *Module) RetrievalRecords() ([]RetrievalRecord, error) {
+	return m.store.getRetrievals()
+}
+
+// Close closes the deals Module.
+func (m *Module) Close() error {
+	return m.store.close()
+}
+
+func (m *Module) initPendingDeals() {
+	pendingDeals, err := m.store.getPendingDeals()
+	if err != nil {
+		log.Errorf("getting pending deals: %v", err)
+		return
+	}
+	for _, dr := range pendingDeals {
+		remaining := time.Until(time.Unix(dr.Time, 0).Add(dealTimeout))
+		if remaining <= 0 {
+			go m.finalizePendingDeal(dr)
+		} else {
+			go m.eventuallyFinalizeDeal(dr, remaining)
+		}
+	}
+}
+
+func (m *Module) recordDeal(params *api.StartDealParams, proposalCid cid.Cid) {
+	di := DealInfo{
+		PieceCID:      params.Data.Root,
+		Duration:      params.MinBlocksDuration,
+		PricePerEpoch: params.EpochPrice.Uint64(),
+		Miner:         params.Miner.String(),
+		ProposalCid:   proposalCid,
+	}
+	record := DealRecord{
+		Addr:     params.Wallet.String(),
+		Time:     time.Now().Unix(),
+		DealInfo: di,
+		Pending:  true,
+	}
+	log.Infof("storing pending deal record for proposal cid: %s", proposalCid.String())
+	if err := m.store.putPendingDeal(record); err != nil {
+		log.Errorf("storing pending deal: %v", err)
+		return
+	}
+	go m.eventuallyFinalizeDeal(record, dealTimeout)
+}
+
+func (m *Module) finalizePendingDeal(dr DealRecord) {
+	deletePending := func() {
+		if err := m.store.deletePendingDeal(dr.DealInfo.ProposalCid); err != nil {
+			log.Errorf("deleting pending deal for proposal cid %s: %v", dr.DealInfo.ProposalCid.String(), err)
+		}
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
+	info, err := m.api.ClientGetDealInfo(ctx, dr.DealInfo.ProposalCid)
+	if err != nil {
+		log.Errorf("getting deal info: %v", err)
+		deletePending()
+		return
+	}
+	if info.State != storagemarket.StorageDealActive {
+		log.Infof("pending deal for proposal cid %s isn't active yet, deleting pending deal", dr.DealInfo.ProposalCid.String())
+		deletePending()
+	} else {
+		di, err := fromLotusDealInfo(ctx, m.api, info)
+		if err != nil {
+			log.Errorf("converting proposal cid %s from lotus deal info: %v", dr.DealInfo.ProposalCid.String(), err)
+			deletePending()
+			return
+		}
+		record := DealRecord{
+			Addr:     dr.Addr,
+			Time:     time.Now().Unix(), // Note: This can be much later in time than the deal actually became active on chain
```

Comment:
Setting the time for completeness, but the active epoch in `di` is more important.